libkbfs/tlf_journal.go

```diff
@@ -809,6 +808,62 @@ func (j *tlfJournal) getJournalStatus() (TLFJournalStatus, error) {
 	}, nil
 }
 
+func (j *tlfJournal) getJournalStatus() (TLFJournalStatus, error) {
+	j.journalLock.RLock()
+	defer j.journalLock.RUnlock()
+	return j.getBaseJournalStatusLocked()
+}
+
+func (j *tlfJournal) getJournalStatusWithPaths(ctx context.Context,
+	config Config, mdOps journalMDOps, blocks *folderBlockOps) (
+	TLFJournalStatus, error) {
+	j.journalLock.RLock()
+	defer j.journalLock.RUnlock()
+	status, err := j.getBaseJournalStatusLocked()
+	if err != nil {
+		return TLFJournalStatus{}, err
+	}
+
+	// Make chains over the entire range to get the unflushed files.
+	// TODO: cache this chains object and update it when new revisions
+	// appear and when revisions are flushed, instead of rebuilding
+	// every time.
+	if status.RevisionEnd != MetadataRevisionUninitialized {
+		var irmds []ImmutableRootMetadata
+		// These GetRange calls end up taking journalLock for reading
```

Comment:
Unfortunately, it's not ok: https://golang.org/pkg/sync/#RWMutex says

```
If a goroutine holds a RWMutex for reading, it must not expect this or any other goroutine to be able to also take the read lock until the first read lock is released. In particular, this prohibits recursive read locking. This is to ensure that the lock eventually becomes available; a blocked Lock call excludes new readers from acquiring the lock.
```

I actually ran into this while messing around with `leveledLock`; my vague memory is that recursive read-locking caused deadlocks to happen under some circumstances.

I don't think we really need to hold this under lock, though, right? In fact, I think a better structure would be to pull this out of `tlfJournal` into `journalServer`, and just have `journalServer` call `getJournalStatus`. Then it would just have to handle `GetRange` returning a partial/empty range.